Домашнее задание

xafish opened the pull request:
hw07
"Белая магия" Spring Data: Spring Data JPA

xafish committed:
Домашнее задание
hw07
"Белая магия" Spring Data: Spring Data JPA

kataus commented on hw07/src/main/java/ru/otus/spring/repositories/BookRepository.java:
```diff
@@ -0,0 +1,29 @@
+    @EntityGraph(attributePaths = {"author", "genre", "comment"})
+    Book findByName(String name);
+
+    /*List<Book> findAll();
```
Мертвый код не храним пока он нас под собой не похоронил

kataus commented on hw07/src/main/java/ru/otus/spring/repositories/BookRepository.java:
```diff
@@ -0,0 +1,29 @@
+
+public interface BookRepository extends JpaRepository<Book, Long> {
+
+    @EntityGraph(attributePaths = {"author", "genre", "comment"})
```
И тут мы вытянули в оперативку всю базу. Не надо лишнего тянуть, это дорого будет

kataus commented on hw07/src/main/java/ru/otus/spring/repositories/CommentRepository.java:
```diff
@@ -0,0 +1,19 @@
+
+public interface CommentRepository extends JpaRepository<Comment, Long> {
+
+    @EntityGraph(attributePaths = {"book", "book.author", "book.genre"})
```
Так уж сразу всё нам надо?

kataus commented on hw07/src/main/java/ru/otus/spring/service/BookLibraryServiceImpl.java:
```diff
@@ -0,0 +1,144 @@
+@Service
+public class BookLibraryServiceImpl implements BookLibraryService{
+
+    @PersistenceContext
```
? Откуда он в сервисе? Этим занимается дао слой

kataus commented on hw07/src/main/java/ru/otus/spring/service/BookLibraryServiceImpl.java:
```diff
@@ -0,0 +1,144 @@
+
+    @Override
+    @Transactional
+    public String addBook(@ShellOption({"id", "u"}) Long id,
```
Зачем намм ShellOption в классе бизнес-логики? UI и бизнес-логика - это разные задачи, которые будут решать разные бины

kataus commented on hw07/src/main/java/ru/otus/spring/domain/Book.java:
```diff
@@ -0,0 +1,64 @@
+    @JoinColumn(name = "id_author")
+    private final Author author;
+
+    @ManyToOne(fetch = FetchType.LAZY)
```
Зачем LAZY, если мы всегда выгружаем сразу? Сделать EAGER и не городить огород с графами

kataus commented on hw07/src/main/java/ru/otus/spring/domain/Comment.java:
```diff
@@ -0,0 +1,44 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private final Long id;
+
+    @ManyToOne(fetch = FetchType.LAZY)
```
Без книги нет пользы от комментария, лучше сразу EAGER сделать и делегировать построение связей Hibernate

xafish committed:
Домашнее задание
hw07
"Белая магия" Spring Data: Spring Data JPA
+ внёс правки по комметариям

kataus commented on hw07/src/main/java/ru/otus/spring/repositories/BookRepository.java:
```diff
@@ -0,0 +1,17 @@
+
+    List<Book> findAll();
+    Optional<Book> findById(Long id);
+    @EntityGraph(attributePaths = {"comment"})
```
Прям нам всегда сразу будут нужны комментарии? Хорошо Optional с findById, почему бы и тут не сделать?

kataus commented on hw07/src/main/java/ru/otus/spring/service/BookLibraryServiceImpl.java:
```diff
@@ -0,0 +1,137 @@
+    private final CommentRepository commentRepository;
+
+    @Override
+    @Transactional
```
Зачем? Не сильно ли круто будет?

kataus commented on hw07/src/main/java/ru/otus/spring/service/BookLibraryServiceImpl.java:
```diff
@@ -0,0 +1,137 @@
+                b-> result.set(result + (bookToString(b) + "\n"))
+        );
+
+        return result.get();
```
Слишком много бизнес-логика знает о нюансах представления. Отдать список книг и не париться, а слой представления уже сделает всё для нас?

xafish committed:
Домашнее задание
hw07
"Белая магия" Spring Data: Spring Data JPA
+ внёс правки по комметариям
+ доработки
